contributing/building-a-bento-amp-extension.md

```diff
@@ -0,0 +1,681 @@
+  on="slideChange:my-selector.toggle(index=e.index, value=true)"
+></amp-base-carousel>
+```
+
+To fire events on your element use AMP's action service and the
+`.trigger` method.
+
+```javascript
+actionServiceForDoc(doc.documentElement).trigger(
+  this.element_,
+  'slideChange',
+  createCustomEvent(
+    win,
+    `amp-base-carousel.${name}`,
+    dict({'index': index})
+  ),
+  ActionTrust.DEFAULT
+);
+```
+
+And to expose actions use `registerApiAction` method that your element
+inherits from `PreactBaseElement`. Note, this should correspond directly with the API exposed in the corresponding Preact compnent via `useImperativeHandle`, and the component should be defined using a `forwardRef` accordingly.
+
+```javascript
+this.registerApiAction('close', this.close.bind(this));
+```
+
+Your element could also choose to override the `activate` method
+inherited from BaseElement that would define the default action for your
+element. For example `amp-lightbox` overrides `activate` to define the `open`
+default case.
+
+Make sure your element documentation documents the events and actions it
+exposes both in its own documentation and in [`AMP Actions and Events`](https://github.com/ampproject/amphtml/blob/master/spec/amp-actions-and-events.md).
+
+## Allowing proper validations
+
+One of AMP's features is that a document can be checked against
+validation rules to confirm it's AMP-valid. While your Bento component can be used in non-AMP contexts, it should also smoothly integrate with AMP documents. When you implement your element, the AMP validator needs to be updated to add rules for your element to keep documents using your element valid. In order to do that you need
+to file an issue on the GitHub repo select "Related to: Validator" and
+mention what rules the validator needs to validate. This usually
+includes
+
+-   Your element tag-name
+-   Required attributes for the element
+-   Specific values that an attribute accepts (e.g. `myattr="TYPE1|TYPE2"`)
+-   Layouts your element supports (see [Layout specs](https://github.com/ampproject/amphtml/blob/master/spec/amp-html-layout.md) and [Layouts supported in your element](#layouts-supported-in-your-element))
+-   If there are restrictions where your element can or can't appear (e.g. disallowed_ancestory, mandatory_parent...)
+
+For more details take a look at [Contributing Component Validator
+Rules](https://github.com/ampproject/amphtml/blob/master/contributing/component-validator-rules.md).
+
+## Performance considerations
+
+### Loading external resources
+
+If your extension needs to load external resources (like an sdk) then
+you might need to add proper third party integration for it to work and
+use the proper third party iframe. Loading external resources is only
+allowed inside a 3p iframe which AMP serves on a different domain for
+security and performance reasons. Take a look at adding
+[&lt;amp-facebook&gt;](https://github.com/ampproject/amphtml/pull/1479/files)
```

Comment:
```suggestion
[`<amp-facebook>`](https://github.com/ampproject/amphtml/pull/1479/files)
```

Reply:
Removed the arrows entirely here, they aren't necessary.